src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +571,656 @@ TODO: check this register,doesn't seem to be 100% correct.
 */
 
 
-void magicard_state::video_start()
+uint32_t magicard_base_state::screen_update_magicard(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
 {
-}
+	uint8_t *src;
+	uint32_t *dest;
+	int scanline = screen.vpos();
 
-uint32_t magicard_state::screen_update_magicard(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
-{
-	// TODO: border & genlock
-	bitmap.fill(m_palette->black_pen(), cliprect);
-
-	 // punt if display enable is off
-	if(!(SCC_DE_VREG))
+	if(scanline < 32)
 		return 0;
 
-	uint32_t count = ((SCC_VSR_VREG) / 2);
-
-	if(SCC_FG_VREG)  // 4bpp gfx
-	{
-		for(int y = 0; y < 300; y++)
-		{
-			for(int x = 0; x < 84; x++)
-			{
-				uint32_t color;
-
-				color = ((m_magicram[count]) & 0x000f) >> 0;
+	dest = &bitmap.pix(scanline);
 
-				if(cliprect.contains((x * 4) + 3, y))
-					bitmap.pix(y, (x * 4) + 3) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0x00f0) >> 4;
-
-				if(cliprect.contains((x * 4) + 2, y))
-					bitmap.pix(y, (x * 4) + 2) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0x0f00) >> 8;
-
-				if(cliprect.contains((x * 4) + 1, y))
-					bitmap.pix(y, (x * 4) + 1) = m_palette->pen(color);
-
-				color = ((m_magicram[count]) & 0xf000) >> 12;
-
-				if(cliprect.contains((x * 4) + 0, y))
-					bitmap.pix(y, (x * 4) + 0) = m_palette->pen(color);
-
-				count++;
-			}
-		}
-	}
-	else  // 8bpp gfx
+	src = m_scc66470->line(scanline);
+	for(int x = 0 ; x < 768 ; x++)
```

Comment:
Same as above, this can be simplified to for(int x = cliprect.min_x; x <= cliprect.max_y ; x++)